src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs

```diff
@@ -1439,10 +1439,13 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
                 // When MaximumRetryCount is not specified, the totalRequests == 1.
                 if (totalRequests > 1 && ShouldRetry(response.StatusCode))
                 {
+                    // If the status code is 429 get the retry interval from the Headers.
+                    WebSession.RetryIntervalInSeconds = ((int)response.StatusCode == 429 && response.Headers.TryGetValues("Retry-After", out var retryAfter)) ? Convert.ToInt32(retryAfter) : RetryIntervalSec;
```

Comment:
I cannot agree with:
1. we rewrite WebSession - if next response will be not 429 we will use value from previous 429 response, I think it is not correct.
2. `out var retryAfter` - we should use explicit type
3. then the type is IEnumerable - I wonder how `Convert.ToInt32(retryAfter)` works, perhaps we should take first element and what if it absent? - should we report or silently use our default?

Reply:
I will initially focus on fixing 2 and 3.

Reply:
I think we should silently use our default

Reply:
2 and 3 should be fixed (it now actually works, I tested it and my last implementation was broken).

Reply:
I have run some tests (`Invoke-WebRequest "https://httpbin.org/status/402"`), removed some code and I think that 1 is not really a problem

Reply:
It is wrong behavior - this header should only affect the current response and not the others.

Reply:
I'm sorry, I wasn't clear in my last comment, I meant that in my opinion the header only affects the current response, I might be wrong. Could you recommend a good way to test a 429 -> 4xx redirect?

Reply:
Mockoon Environment (zipped because json is not supported on github): [test.zip](https://github.com/PowerShell/PowerShell/files/10147274/test.zip)

Mockoon gui:
![mockoon](https://user-images.githubusercontent.com/105941898/205459994-8cfaccd7-17e3-423e-8094-c80656558aba.png)

- I modified WebRequestPSCmdlet.Common adding `WriteVerbose(((int)response.StatusCode).ToString());` on the line before the ternary operator to show the StatusCode.

- I created a Mockon Environment that randomly alternates a 429 http response (with Retry-After=2) and a 400 http response.

- I tested the mock api.

```ps1
Invoke-WebRequest "http://localhost:3000/429-test" -Verbose -SkipHttpErrorCheck -MaximumRetryCount 10 
```

Results:

![429](https://user-images.githubusercontent.com/105941898/205460033-7866723c-dad2-48ca-96ff-e45ddcdda499.png)

It seems like everything works as intended, what do you think @iSazonov ?

Reply:
I don't know how your test work but we can create WebSession object before and then use it is some iwr. So if we change WebSession object in first iwr call we will use the bad one in follow calls.

Reply:
I tested it once again following your suggestion:

- 429 endpoint with Retry-After=2
- 403 endpoint with no headers

![2](https://user-images.githubusercontent.com/105941898/205609829-693d825d-38f0-4a3b-9a4b-8f20d05a895e.png)

In my opinion it works correctly.

The `$session` object leaks the last `RetryIntervallInSeconds`, but that's also the behaviour of Powershell 7.3.

![3](https://user-images.githubusercontent.com/105941898/205612276-3631d1de-1c40-4fec-8007-68a599ef9540.png)
![4](https://user-images.githubusercontent.com/105941898/205612287-43dcc04c-4def-4e30-8512-71eceabf2337.png)

We could reset `RetryIntervallInSeconds` to 0 or the default after the request if this is a problem.

Reply:
> The $session object leaks the last RetryIntervallInSeconds, but that's also the behaviour of Powershell 7.3.

Can you explain/point line in original code?

Reply:
`WebSession.RetryIntervalInSeconds` is defined on line 694 of the original code, so when `MaximumRetryCount` > 0 the session variable always reports the last `RetryIntervallInSeconds`. It gets replaced by the default (5) or by the number set by the user with `-RetryIntervalSec` on the following web requests.

```cs
            if (MaximumRetryCount > 0)
            {
                WebSession.MaximumRetryCount = MaximumRetryCount;

                // only set retry interval if retry count is set.
                WebSession.RetryIntervalInSeconds = RetryIntervalSec;
            }
        }
```
Simple test:

![5](https://user-images.githubusercontent.com/105941898/205639156-4eee0013-b0a0-4a18-9193-0cfd75c0542e.png)

Reply:
So if we look WebSession properties after first call we see a value which will be overwrited on next call. It is too tricky, unreliable and can mislead. We should safe a value in WebSession only if we use it later.

Reply:
In line 1443 we could define local variable instead of `WebSession.RetryIntervalInSeconds`.

Reply:
Done

Reply:
> In line 1443 we could define local variable instead of `WebSession.RetryIntervalInSeconds`.

Perhaps we could do the same thing for `WebSession.MaximumRedirection` in another PR, what do you think?

Reply:
Yes, let's do this in follow PR.

Reply:
It seems I mislead you. It is not right intention.

Reply:
I will close the other PR